src/lib/protocols/http.c

```diff
@@ -420,19 +420,44 @@ static void ndpi_http_parse_subprotocol(struct ndpi_detection_module_struct *ndp
 
 static void ndpi_check_user_agent(struct ndpi_detection_module_struct *ndpi_struct,
 				  struct ndpi_flow_struct *flow,
-				  char *ua) {
-  u_int len;
+				  char const *ua, size_t ua_len) {
   char *double_slash;
   
   if((!ua) || (ua[0] == '\0'))
     return;
-  else
-    len = strlen(ua);
+
+  if (ua_len > 12)
+  {
+    size_t i, upper_case_count = 0;
+
+    for (i = 0; i < ua_len; ++i)
+    {
+      /*
+       * We assume at least one non alpha char.
+       * e.g. ' ', '-' or ';' ...
+       */
+      if (isalpha(ua[i]) == 0)
+      {
+        break;
+      }
+      if (isupper(ua[i]) != 0)
+      {
+        upper_case_count++;
+      }
+    }
+
+    if (i == ua_len)
+    {
+      float upper_case_ratio = (float)upper_case_count / (float)ua_len;
+      if (upper_case_ratio >= 0.2f)
```

Comment:
Empiric value?

Reply:
observation of some malware traffic, but not yet empiric